Update embedsignature directive to emit function annotations

dalcinl opened the pull request:
This is for sure not enough, but at least it can handle "simple" annotations like non-compound compile-time constants and name nodes.

scoder commented:
Should we start using the `DeclarationWriter` in `Cython.CodeWriter` for the expressions? I can't find any current usage, but a `TreeVisitor` seems the right approach to me.

We could extract an `ExpressionWriter` base class from it that excludes the declaration parts and could be used wherever we want to generate a text representation of Cython expressions.

dalcinl commented:
@scoder I would need some help with that, I would like to get this working for the next release, but I'm a bit lost about how to get started.

scoder commented:
It's used like this in `TestUtils.py`:
```
    def codeToLines(self, tree):
        writer = CodeWriter()
        writer.write(tree)
        return writer.result.lines
```
I would extract all `visit_*()` methods into the new base class (inheriting from `TreeVisitor`) that handle nodes allowed in a Cython expression, and then add new methods for nodes that are currently unsupported. Look at what's there already (`NodeNode`, `IntNode`, etc.), most nodes should be really straight forward.

dalcinl commented:
@scoder Thanks! That helped a lot. I'll give it a look tomorrow.

dalcinl commented:
@scoder PR updated. After a some thinking, it is not obvious to me that `DeclarationWritter` has much to share with the new `ExpressionWriter` I implemented, so I suggest to keep them separate for now. Please note the change in the testcase, are you OK with that? Otherwise, could you suggest a proper fix?

scoder commented on Cython/CodeWriter.py:
```diff
@@ -519,3 +519,276 @@ def visit_CFuncDefNode(self, node):
+    }
+
+    def operator_enter(self, new_prec):
+        if not hasattr(self, 'precedence'):
```
This looks like there should be a default value for `precedence` in the class.

dalcinl replied:
Yea, you are definitely right...

scoder commented on tests/run/embedsignatures.pyx:
```diff
@@ -185,7 +188,7 @@ __doc__ = ur"""
 
     >>> print(funcdoc(f_defexpr5))
-    f_defexpr5(int x=4)
+    f_defexpr5(int x=2 + 2)
```
This change has pros and cons. If it's really spelled that way in the source file, then it's probably nice to spell it the same way in the docs. However, the expression might stem from a compile time calculated value, which the programmer might not have wanted to leak into the docs. Unclear. I think there might be a case for collapsing the expressions of default arguments and keeping the expressions of annotations.

dalcinl replied:
Well, I guess this is not a blocker for this PR. We can decide later on it, right?

scoder commented on Cython/CodeWriter.py:
```diff
@@ -519,3 +519,276 @@ def visit_CFuncDefNode(self, node):
+        return self.result
+
+    def put(self, s):
+        self.result += s
```
Wouldn't it be better if this appended to a list, and write() joined and returned it?

dalcinl replied:
Is that a question or a request? Are you concerned about performance?

dalcinl replied:
@scoder IIUC, any python expression has to be written in a single line (unless parentized). I still don't see the point of appending to a list to join at the end.

scoder replied:
Just commenting. I guess it's ok for small stuff, but I'd always go with a list in such a case since you never know what people end up using it for.

dalcinl replied:
Note that `DeclarationWriter` is doing basically the same for each line. I would propose to defer a final decision on this to a future refactoring of `DeclarationWriter`/`ExpressionWriter`.

scoder commented on Cython/CodeWriter.py:
```diff
@@ -519,3 +519,276 @@ def visit_CFuncDefNode(self, node):
+    def visit_BinopNode(self, node):
+        op = node.operator
+        prec = self.binop_precedence.get(op, 0)
+        self.operator_enter(prec)
```
Nice.

scoder commented on Cython/CodeWriter.py:
```diff
@@ -519,3 +519,276 @@ def visit_CFuncDefNode(self, node):
+        self.emit_sequence(node, u"()")
+
+    def visit_SetNode(self, node):
+        self.emit_sequence(node, u"{}")
```
Except for the empty `SetNode`, which can get inserted for `set()`.

dalcinl replied:
Good catch!

scoder commented on Cython/CodeWriter.py:
```diff
@@ -519,3 +519,276 @@ def visit_CFuncDefNode(self, node):
+            repr_val = repr_val[1:]
+        self.put(u"%s" % repr_val)
+
+    def visit_UnicodeNode(self, node):
```
These three could potentially be joined by always lstripping `"ub"`.

dalcinl commented:
@scoder I think I've addressed most of your concerns. I'm still not sure about `DeclarationWriter` inheriting from `ExpressionWriter`, but such refactor should be easy to implement in a future PR.

scoder commented:
Thanks! I'll let travis finish its run and then happily merge it.
